ODC-4981: Add new plugin to support Shipwright Builds and BuildRuns

jerolimov opened the pull request:
**Fixes**: 
* Epic [ODC-4981 Support Builds v2 in Dev Console](https://issues.redhat.com/browse/ODC-4981), incl.
  * [ODC-6170 Script to install operator and create a list of different build states](https://issues.redhat.com/browse/ODC-6170)
  * [ODC-6171 Create shipwright-plugin and routes to add custom pages](https://issues.redhat.com/browse/ODC-6171)
  * [ODC-6172 Shipwright Build detail page Create Shipwright Build detail page](https://issues.redhat.com/browse/ODC-6172)
  * [ODC-6172 Shipwright BuildRun detail page](https://issues.redhat.com/browse/ODC-6172)
  * [ODC-6174 Shipwright Build and BuildRun list](https://issues.redhat.com/browse/ODC-6174)

**Solution Description**: 
Added a new static plugin `shipwright-plugin`

1. Custom list for Build and BuildRuns
2. Custom details pages for Build and BuildRuns
3. Both shows the BuildSpec on the right side, and have a Events tab
5. Builds has also a tab to list or related BuildRuns
6. BuildRuns has also a tab to view the TaskRun log
7. Option to start a Build and rerun a BuildRun (without any option to configure some parameters)

**Screen shots / Gifs for design review**: 

### Navigation

Admin perspective:

![admin](https://user-images.githubusercontent.com/139310/172580867-532ce036-a446-450c-bd08-bfdd0f5ce319.gif)

Dev perspective:

![dev](https://user-images.githubusercontent.com/139310/172580996-545f43aa-8e9d-484f-a247-c944670ae0bb.gif)

### Shipwright Builds

Shipwright Builds list:

![image](https://user-images.githubusercontent.com/139310/172581278-834bb825-a3f0-44b6-afdf-99c48ee23192.png)

Shipwright Build Details tab:

![image](https://user-images.githubusercontent.com/139310/172581341-b6c8403b-1b01-4cdf-9588-2807f5054c82.png)

Shipwright Build BuildRuns tab:

![image](https://user-images.githubusercontent.com/139310/172581410-11e177f3-1b94-4429-afa2-2472aee44a9c.png)

Shipwright Build Events tab:

![image](https://user-images.githubusercontent.com/139310/172581564-ace70e71-0c79-44b7-acd8-3228439e0f3e.png)

### Shipwright BuildRuns

Shipwright BuildRun list:

![image](https://user-images.githubusercontent.com/139310/172581819-bc17f86d-8e35-4eaa-a1c7-4c516111145d.png)

Allow you to take a quick peek into the error:

![image](https://user-images.githubusercontent.com/139310/172581911-a9ba52d0-c418-421a-a039-32682f7d78fa.png)

Shipwright BuildRun Details tab:

![image](https://user-images.githubusercontent.com/139310/172582001-84408bbb-f9c2-40dc-8e9a-6be91fcb174c.png)

Shipwright BuildRun Logs tab (similar aka reuses a TaskRun log viewer):

![image](https://user-images.githubusercontent.com/139310/172582087-bc1c144b-27b3-467e-b9f9-77f41944b6fd.png)

Shipwright BuildRun Events tab:

![image](https://user-images.githubusercontent.com/139310/172582271-73790857-e5f6-47db-84af-dde975b7a945.png)

**Unit test coverage report**: 
Added some, more will be added in another PR.

**Test setup:**

Tested with version 0.9

## Installation

1. Install `Red Hat OpenShift Pipelines` from the OperatorHub (Red Hat catalog)
8. Install `Shipwright Operator` from the OperatorHub (operatorhubio-catalog)
9. On the installed operator details page, select `ShipwrightBuild` and create the default resource.

   This will enable the Shipwright controller and install the CRDs (`ClusterBuildStrategy`, `BuildStrategy`, `Build`, `BuildRun` etc.)

## Test builds
Tested with Shipwright operator 0.9.

1. The origin samples in the Shipwright git repository doesn't run with the default Service Account on OpenShift and needs some additional configurations.
2. Our samples are slightly modified and doesn't require this. We use namespaced scoped BuildStrategies so that they don't conflict.
3. We also uses our internal OpenShift image repository and provided some full demos.

To install the resources you need to create the namespace `build-examples` and apply the YAML files from the included samples folder.

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

/cc @invincibleJai @divyanshiGupta

invincibleJai commented on frontend/packages/shipwright-plugin/src/actions/useBuildActions.ts:
```diff
@@ -0,0 +1,40 @@
+              errorModal({ error: err.message });
+            });
+        },
+        // accessReview: asAccessReview(BuildModel, obj, 'create'),
```
??

jerolimov replied:
Thanks! Added missing accessReview:

```tsx
      accessReview: {
        verb: 'create',
        group: BuildRunModel.apiGroup,
        resource: BuildRunModel.plural,
        namespace: buildRun.metadata?.namespace,
      },
```

invincibleJai commented on frontend/packages/shipwright-plugin/src/api.ts:
```diff
@@ -0,0 +1,63 @@
+      kind: 'BuildRun',
+      metadata: {
+        namespace: buildRun.metadata.namespace,
+        generateName: `${buildName}-`,
```
why do we have a suffix as `-` on `generateName`?

divyanshiGupta commented on frontend/packages/shipwright-plugin/src/actions/useBuildActions.ts:
```diff
@@ -0,0 +1,40 @@
+import { startBuild } from '../api';
+import { Build } from '../types';
+
+export default function useBuildActions(build: Build) {
```
At @jerolimov this is also fine but to maintain code consistency you can use change it to an arrow function.
```suggestion
const useBuildActions = (build: Build) => {
```

divyanshiGupta commented on frontend/packages/shipwright-plugin/src/actions/useBuildActions.ts:
```diff
@@ -0,0 +1,40 @@
+  );
+
+  return [actions, !inFlight, undefined];
+}
```
```suggestion
}
export default useBuildActions;
```

jerolimov replied:
:+1: changed. will be part of the next commit.

divyanshiGupta commented on frontend/packages/shipwright-plugin/src/actions/useBuildRunActions.ts:
```diff
@@ -0,0 +1,45 @@
+import { canRerunBuildRun, rerunBuildRun } from '../api';
+import { BuildRun } from '../types';
+
+export default function useBuildRunActions(buildRun: BuildRun) {
```
Same as above

jerolimov replied:
done

divyanshiGupta commented on frontend/packages/shipwright-plugin/src/components/build-details/BuildDetailsTab.tsx:
```diff
@@ -0,0 +1,30 @@
+  return (
+    <div className="co-m-pane__body">
+      <SectionHeading text={t('shipwright-plugin~Build details')} />
+
```
```suggestion
```

jerolimov commented on frontend/packages/dev-console/src/components/builds/BuildsTabListPage.tsx:
```diff
@@ -0,0 +1,122 @@
+import CreateProjectListPage from '../projects/CreateProjectListPage';
+
+interface ServingListPageProps {
+  match: Rmatch<{ ns: string }>;
```
:+1: done

jerolimov commented on frontend/packages/dev-console/src/components/builds/BuildsTabListPage.tsx:
```diff
@@ -0,0 +1,122 @@
+  // We might add a generic extension for multi tab list pages in the future.
+  // Currently we just check if Shipwright is installed and load a (dynamic) resource list page.
+  const shipwrightBuildSupported = useFlag('SHIPWRIGHT_BUILD');
+  const shipwrightBuildLoader = resourceListPages.get('shipwright.io~v1alpha1~Build');
+  const [shipwrightBuildModel] = useK8sModel('shipwright.io~v1alpha1~Build');
```
Don't want to add a dependency from `dev-console` to `shipwright-plugin` for this. The even better solution is a generic extension that allows us to define "Builds" and then implement this in both packages. Wdyt?

invincibleJai replied:
yeah, that will make sense to add a tech debt for it and handle it.

jerolimov replied:
Created this ticket for this: https://issues.redhat.com/browse/ODC-6693

divyanshiGupta commented on frontend/packages/shipwright-plugin/src/components/buildrun-list/BuildRunTable.tsx:
```diff
@@ -0,0 +1,123 @@
+
+export const BuildRunHeader = () => {
+  // TODO: Create bug so that a header can be written as component?
+  // const { t } = useTranslation();
```
Remove comment?

jerolimov commented on frontend/packages/shipwright-plugin/samples.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
removed (with next commit)

invincibleJai commented on frontend/packages/shipwright-plugin/src/actions/useBuildRunActions.ts:
```diff
@@ -0,0 +1,45 @@
+                    errorModal({ error });
+                  });
+              },
+              // accessReview: asAccessReview(BuildRunModel, obj, 'create'),
```
same as above, need to add access check

divyanshiGupta commented on frontend/packages/dev-console/src/components/builds/BuildsTabListPage.tsx:
```diff
@@ -0,0 +1,122 @@
+    });
+  }
+
+  if (!match.params.ns) {
```
```suggestion
  if (!namespace) {
```

divyanshiGupta commented on frontend/packages/dev-console/src/components/builds/BuildsTabListPage.tsx:
```diff
@@ -0,0 +1,122 @@
+          <Trans t={t} ns="devconsole">
+            Select a Project to view the list of builds or{' '}
+            <Button isInline variant="link" onClick={openProjectModal}>
+              create a Project
```
```suggestion
              Create a Project
```

jerolimov replied:
Used the same text from https://github.com/openshift/console/blob/release-4.10/frontend/packages/dev-console/src/components/buildconfig/BuildConfigPage.tsx#L35 and just changed `list of BuildConfigs` to `list of builds` as this can show the other Builds now as well.

divyanshiGupta commented on frontend/packages/dev-console/src/components/builds/BuildsTabListPage.tsx:
```diff
@@ -0,0 +1,122 @@
+        pages={pages}
+        match={match}
+        menuActions={menuActions}
+        // createRedirect
```
Remove comment?

jerolimov commented on frontend/packages/shipwright-plugin/src/components/build-list/BuildTable.tsx:
```diff
@@ -0,0 +1,101 @@
+
+export const BuildHeader = () => {
+  // TODO: Create bug so that a header can be written as component?
+  // const { t } = useTranslation();
```
Updated at least to:

```
  // This function is NOT called as component, so we can not use useTranslation here.
```

jerolimov commented on frontend/packages/dev-console/src/components/builds/BuildsTabListPage.tsx:
```diff
@@ -0,0 +1,122 @@
+  if (shipwrightBuildSupported && shipwrightBuildLoader) {
+    menuActions.shipwrightBuild = {
+      label: t('devconsole~Shipwright Build'),
+      onSelection: () => `/k8s/ns/${namespace}/shipwright.io~v1alpha1~Build/~new`,
```
Added a condition

jerolimov commented:
> @invincibleJai:
> 1. nav option `Builds` get's selected even on logs or builds but on click shows buildConfigs, is this expected (happens only for k8s build)?

That's like before and from my perspective, it's expected and a feature. It's more that we should do this also for Shipwright Builds/BuildRuns maybe in the future.

> 2. Buildrun details don't sow logs or events and have inline text

Implemented this pages yesterday.

> 3. should the tech preview badge be on the left or on right?

This is the default behavior of a `ListPage` without a title and create button. If we want to change this, we should update the general ListPage component. So from my perspective this is unrelated to this PR.

> 4. Filter refreshes as I type

Thanks for this finding! Fixed this today by memorize the AsyncComponent in `BuildsTabListPage`.

jerolimov commented:
PTAL @sanketpathak @serenamarie125 

@serenamarie125 If some label updates are needed, we would like to add this in a separate PR. Please let me, plus Jai or Divyanshi know if there is anything to update. Both because I'm less available the rest of this and next week.

jerolimov commented:
/retest

jerolimov commented:
/test all

jerolimov commented:
/test all

jerolimov commented:
Propagate from the Epic
/label docs-approved
/label px-approved

jerolimov commented:
Tested this on a cluster-bot instance (`launch openshift/console#11641`)

Without the operator installed:

https://user-images.githubusercontent.com/139310/172613302-2cec10d1-75d1-4d25-b61f-ee600dd8fdb5.mp4

Installing all resources:

https://user-images.githubusercontent.com/139310/172613311-4d6f5a7f-30fb-41f3-b43e-fcc0128635b2.mp4

Showing the Shipwright builds:

https://user-images.githubusercontent.com/139310/172613328-6bbb4de3-6def-4383-8786-ee2c834300ff.mp4

/label qe-approved

jerolimov commented:
/retest

invincibleJai commented:
Verified the changes and looks good

/lgtm
